Added IOS XR support for snmp_user

jonnytdevops opened the pull request:
This adds support to IOS XR for snmp_user.
This provider does not support providing unencrypted passwords due to the way salting works on this platform. As such, the minitests that test for equivalency between unencrypted and encrypted passwords have been skipped.

Needs https://github.com/cisco/cisco-network-puppet-module/pull/353/files

```
TestSnmpUser#test_snmpuser_create_destroy_with_engine_id = 7.25 s = .
TestSnmpUser#test_snmpuser_privpassword = 22.61 s = .
TestSnmpUser#test_snmpuser_create_invalid_args = 5.53 s = .
TestSnmpUser#test_snmpuser_auth_password_equal_localizedkey = 5.83 s = .
TestSnmpUser#test_create_1_group_auth_sha_nopriv_pw_localized_localizedkey_false = 5.59 s = .
TestSnmpUser#test_snmpuser_auth_priv_des_password_equal = 5.41 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_sha_nopriv = 5.20 s = .
TestSnmpUser#test_snmpuser_authpassword_with_engineid = 5.08 s = .
TestSnmpUser#test_snmpuser_auth_priv_password_equal_localizedkey = 4.88 s = .
TestSnmpUser#test_snmpuser_create_invalid_cli = 5.29 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_noauth_nopriv = 16.65 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_sha_priv_aes128_pw_localized = 17.27 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_priv_sha_pw_localized = 17.58 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_priv_aes128_pw_localized = 18.79 s = .
TestSnmpUser#test_snmpuser_default_priv_protocol = 4.96 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_nopriv = 4.76 s = .
TestSnmpUser#test_snmpuser_auth_password_not_equal = 5.16 s = .
TestSnmpUser#test_snmpuser_auth_priv_password_equal_empty = 6.43 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_nopriv_pw_localized = 18.55 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_priv_des = 5.00 s = .
TestSnmpUser#test_snmpuser_create_with_multi_group_noauth_nopriv = 5.22 s = .
TestSnmpUser#test_snmpuser_default_priv_password = 5.86 s = .
TestSnmpUser#test_snmpuser_default_auth_protocol = 5.16 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_priv_aes128 = 5.80 s = .
TestSnmpUser#test_snmpuser_privpassword_with_engineid = 5.23 s = .
TestSnmpUser#test_snmpuser_priv_password_equal_with_engineid = 5.11 s = .
TestSnmpUser#test_snmpuser_auth_password_equal = 5.90 s = .
TestSnmpUser#test_snmpuser_authpassword = 15.89 s = .
TestSnmpUser#test_snmpuser_auth_password_equal_invalid_param = 6.36 s = .
TestSnmpUser#test_snmpuser_auth_priv_password_equal_invalid_param = 5.74 s = .
TestSnmpUser#test_snmpuser_auth_password_equal_with_engineid = 5.59 s = .
TestSnmpUser#test_engine_id_valid_and_none = 5.67 s = .
TestSnmpUser#test_snmpuser_auth_password_equal_priv_invalid_param = 5.99 s = .
TestSnmpUser#test_snmpuser_collection_not_empty = 18.44 s = .
TestSnmpUser#test_snmpuser_default_auth_password = 5.49 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_sha_priv_des = 5.38 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_sha_nopriv_pw_localized = 17.66 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_sha_priv_aes128 = 5.78 s = .
TestSnmpUser#test_snmpuser_create_with_single_group_auth_md5_priv_des_pw_localized = 18.19 s = .
TestSnmpUser#test_snmpuser_default_groups = 5.73 s = .
TestSnmpUser#test_snmpuser_destroy = 25.15 s = .

Finished in 380.284039s, 0.1078 runs/s, 0.3313 assertions/s.

41 runs, 126 assertions, 0 failures, 0 errors, 0 skips
```

jonnytdevops committed:
Added IOS XR support for snmp_user

chburket commented on lib/cisco_node_utils/cmd_ref/snmp_user.yaml:
```diff
@@ -1,49 +1,67 @@
-_exclude: [ios_xr]
-
 _template:
   multiple: true
```
consider adding the get_commands to the template, since they seem to be the same in most of the attributes (just override the one or two that are different)?

chburket commented on lib/cisco_node_utils/snmpuser.rb:
```diff
@@ -60,6 +72,15 @@ def initialize_validator(name, groups, authproto, authpass, privproto,
+
+      if platform == :ios_xr && instantiate
+        fail TypeError \
+          unless version.is_a?(Symbol) && [:v1, :v2c, :v3].include?(version)
```
Wouldn't just `unless [:v1, :v2c, :v3].include?(version)` work?

chburket commented on lib/cisco_node_utils/snmpuser.rb:
```diff
@@ -123,10 +148,30 @@ def destroy
+                   engineId:     engineidstr)
+      end
+
+      SnmpUser.users.delete(@name + ' ' + @engine_id) if platform != :ios_xr
```
Move this into the `else` block above?

jonnytdevops committed:
Refactored snmp_user

chburket commented on tests/ciscotest.rb:
```diff
@@ -100,7 +100,7 @@ def platform
   def config_and_warn_on_match(warn_match, *args)
     if node.client.platform == :ios_xr
-      result = super(warn_match, *args, 'commit best-effort')
+      result = super(warn_match, *args, 'commit')
```
I tried to remove "best-effort" back in March, and Glenn Matthews said that there was a reason it was needed, so I left it in.  Not sure if it's still needed or not.

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -92,6 +92,8 @@ def test_snmpuser_create_invalid_args
   end
 
   def test_snmpuser_create_invalid_cli
+    return if platform == :ios_xr
```
Please explicitly skip for xr instead of just returning.

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -113,6 +115,8 @@ def test_snmpuser_create_invalid_cli
   end
 
   def test_engine_id_valid_and_none
+    return if platform == :ios_xr
```
skip

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -143,21 +147,32 @@ def test_snmpuser_create_with_single_group_noauth_nopriv
   end
 
   def test_snmpuser_create_with_multi_group_noauth_nopriv
+    return if platform == :ios_xr
```
skip

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -179,12 +200,19 @@ def test_snmpuser_destroy
   end
 
   def test_snmpuser_auth_password_equal_invalid_param
+    return if platform == :ios_xr
```
skip

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -194,6 +222,8 @@ def test_snmpuser_auth_password_equal_invalid_param
   end
 
   def test_snmpuser_auth_priv_password_equal_invalid_param
+    return if platform == :ios_xr
```
skip

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -205,6 +235,8 @@ def test_snmpuser_auth_priv_password_equal_invalid_param
   end
 
   def test_snmpuser_auth_password_equal_priv_invalid_param
+    return if platform == :ios_xr
```
skip here and the following tests that return immediately on XR

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -282,7 +326,11 @@ def test_snmpuser_auth_priv_des_password_equal
     assert(snmpuser.priv_password_equal?(priv_pw, false))
   end
 
+  ############################ HERE HERE HERE ########################################
```
remove?

chburket commented on tests/test_snmpuser.rb:
```diff
@@ -292,31 +340,45 @@ def test_snmpuser_create_with_single_group_auth_md5_nopriv
     assert_show_match(
       pattern: /#{user_pat(name)} auth md5 \S+ localizedkey/,
-      command: "show run snmp all | in #{name} | no-more")
+      command: 'show run snmp all | in #{name} | no-more')
```
why single quotes here?

jonnytdevops committed:
Updated test_snmpuser.rb

chrisvanheuveln commented on tests/test_snmpuser.rb:
```diff
@@ -59,11 +59,11 @@ def teardown
     assert_empty(delta, 'Users not deleted after test!')
   end
 
```
fyi, we've done some cleanups in the test files this week that may cause some merge conflicts for you. It's mostly just method name changes to make them shorter. test_snmpuser diffs here:  https://github.com/cisco/cisco-network-node-utils/commit/b91c16854a612a2dc5d522a6a7cda6dbb0ef5dd6

This was done simply to make the test results more readable in the regression reports.

jonnytdevops committed:
Fixed NXOS issue with snmp_user

jonnytdevops committed:
Updated test_snmpuser.rb for NXOS

jonnytdevops commented:
@mikewiebe I have ran the tests on our NXOS image. After making some changes (which have been pushed), all tests pass except those that require des56 priv mode, which isn't supported on the version that we have here. Since these tests haven't been altered by us, I'm assuming that it's our version of N9K that we have.

@chburket I believe I've addressed all comments, thanks

jonnytdevops commented:
@chburket @chrisvanheuveln Can this be merged now please? The puppet side has already been merged. Thanks
